src/QtiItems/QtiItemAssetCloudFrontReplacer.php

```diff
@@ -0,0 +1,164 @@
+{
+    /**
+     * The host for the external source
+     */
+    const OPTION_HOST = 'host';
+
+    /**
+     * The prefix for the path
+     */
+    const OPTION_PREFIX = 'prefix';
+
+    /**
+     * {@inheritDoc}
+     */
+    public function shouldBeReplaced(PackedAsset $packetAsset): bool
+    {
+        return !$this->isExcluded($this->getFilenameFormPacket($packetAsset));
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @throws \common_Exception
+     */
+    public function replace(PackedAsset $packetAsset, string $itemId): PackedAsset
+    {
+        $filename = $this->getFilenameFormPacket($packetAsset);
+
+        $awsClient = $this->getAwsClient();
+        $s3Client = $awsClient->getS3Client();
+        $config = new Config();
+
+        $itemAssetReplacement = $this->getItemAssetsReplacement();
+
+        if (!($itemAssetReplacement instanceof ItemCloudFrontReplacement)) {
+            throw new \common_Exception('The ItemCloudFrontReplacement service should be configured properly for using CloudFront assets.');
+        }
+
+        $s3Adapter = $this->getAwsS3Adapter(
+            $s3Client,
+            $itemAssetReplacement->getOption(ItemCloudFrontReplacement::OPTION_S3_BUCKET),
+            $this->getOption(self::OPTION_PREFIX)
+        );
+
+        $path = $this->buildPath($itemId);
+
+        $s3Adapter->writeStream($path . $filename, $this->getResourceFromPacket($packetAsset), $config);
+        $path = $this->getOption(self::OPTION_HOST) . DIRECTORY_SEPARATOR . $path . $filename;
+        $packetAsset->setReplacedBy($path);
+
+        return $packetAsset;
+    }
+
+    /**
+     * Build a final path for an asset
+     */
+    private function buildPath(string $itemId): string
+    {
+        return 'items' . DIRECTORY_SEPARATOR . $itemId . DIRECTORY_SEPARATOR . 'assets' . DIRECTORY_SEPARATOR ;
+    }
+
+    /**
+     * Get a filename from PacketAsset
+     */
+    private function getFilenameFormPacket(PackedAsset $packetAsset): string
+    {
+        $mediaAsset = $packetAsset->getMediaAsset();
+        $identifier = $mediaAsset->getMediaIdentifier();
+        return $mediaAsset->getMediaSource()->getBaseName($identifier);
+    }
+
+    /**
+     * Get resource from PacketAsset
+     */
+    private function getResourceFromPacket(PackedAsset $packetAsset)
+    {
+        $mediaAsset = $packetAsset->getMediaAsset();
+        $identifier = $mediaAsset->getMediaIdentifier();
+        $fileStream = $mediaAsset->getMediaSource()->getFileStream($identifier);
+        return $fileStream->detach();
+    }
+
+    /**
+     * Check if an asset should be excluded from the moving to the CloudFront
+     */
+    private function isExcluded(string $src): bool
+    {
+        if ($this->hasOption(self::OPTION_EXCLUDE_PATTERNS)) {
+            return $this->checkPatterns($src, $this->getOption(self::OPTION_EXCLUDE_PATTERNS));
+        }
+        return $this->checkPatterns($src, self::EXCLUDE_PATTERNS);
+    }
+
+    /**
```

Comment:
I think the phpdoc is not necessary for `private` methods, parameter declarations are already supported by `php` and the method name should describe fully what it does, not a phpdoc comment